### PR TITLE
Add extension function for Filename

### PR DESCRIPTION
**Pull request by fxfactorial**
@alainfrisch continuing from https://github.com/ocaml/ocaml/pull/609 I couldn't reopen that one because I forced pushed and github greyed out the reopen pull request button. 

I think this is now correct.

### CODE REVIEW

**Comment by alainfrisch** [comment]
I'm concerned by the lack of coherence between this new function and `chop_extension`, which fails when there is no `.` in the basename.  I don't see how to justify this difference.

I would not object to changing the behavior of `chop_extension` so that it returns its argument instead of failing, but this might in theory break code that rely on the current behavior.

Btw, `chop_extension` gives a nice implementation technique for `extension`.  The proposed implementation relies on `basename`, whose semantics is not completely straightforward and leads to suprising results.  For instance, `basename ""` is `"."`, so applying `extension` to the empty string would return `"."`.

File: stdlib/filename.mli [hunk 1]
```diff
@@ -140,4 +140,9 @@ val quote : string -> string
     Warning: under Windows, the output is only suitable for use
     with programs that follow the standard Windows quoting
     conventions.
- *)
+*)
+
+val extension : string -> string
+(** Get the extension of a file if there is one, includes the . when
```
Comment (hakuch) [inline stdlib/filename.mli:146]: This sentence is not grammatically correct. Replace "includes" with "including".

**Comment by damiendoligez** [comment]
I agree with Alain: you should factor out the `search_dot` loop from `chop_extension` into an auxiliary function, then use that in `extension`, thus ensuring that the notion of extension is always the same for the two functions.

Also for consistency, you should raise `Invalid_arg` if there is no extension.

Finally, put the new function right next to `chop_extension` in both `filename.mli` and `filename.ml`.

**Comment by nojb** [comment]
While we are at it, could we also add `Location.absolute_path` ~~(not currently exposed)~~ ?

https://github.com/ocaml/ocaml/blob/trunk/parsing/location.ml#L225-L237

It is handy and not quite a one-liner.

**Comment by damiendoligez** [comment]
> While we are at it, could we also add `Location.absolute_path`

That function has a subtle bug that needs to be fixed first. I don't think we should mix it with this PR.

**Comment by dbuenzli** [comment]
It seems to me that the function as implemented will return `".ocamlinit"` on `".ocamlinit"` which is not what you want if you want to have something that is not broken in the standard library.

You may want to check the preamble [here](http://erratique.ch/software/fpath/doc/Fpath.html#file_exts) for a reasonable and more useful definition of a file extension. Here are a few useful [examples](http://erratique.ch/software/fpath/doc/Fpath.html#ex_get_ext).

**Comment by alainfrisch** [comment]
@dbuenzli I can see the benefits of your definition of extension, but it's certainly a bad idea to have a different definition of the extension for `Filename.extension` and `Filename.chop_extension`.  So the real question is whether we should adopt your definition, including for `Filename.chop_extension`, at the risk of breaking existing code.

**Comment by damiendoligez** [comment]
The current version of `Filename.chop_extension` returns the empty string when applied to `".ocamlinit"`. I don't think anyone ever wants that, so I'd like to take the opportunity to get it right, along with the tiny risk of breaking something with a slight change to the existing function.

**Comment by alainfrisch** [comment]
@damiendoligez I guess one should stick to working on the _last_ extension, right? (i.e. the extension of foo.tar.gz is .gz)

**Comment by alainfrisch** [comment]
I've submitted an implementation of Filename.extension as #669.  It corresponds to @dbuenzli's definition, except when the name ends with a directory separator (the user can apply  `Filename.basename` first if needed).

**Pull request closed without merging**